src/include/cephfs/libcephfs.h

```diff
@@ -1572,6 +1572,51 @@ int ceph_ll_getlk(struct ceph_mount_info *cmount,
 int ceph_ll_setlk(struct ceph_mount_info *cmount,
 		  Fh *fh, struct flock *fl, uint64_t owner, int sleep);
 
+/*
+ * Delegation support
+ *
+ * Delegations are way for an application to request exclusive or
+ * semi-exclusive access to an Inode. The client requests the delegation and
+ * if it's successful it can reliably cache file data and metadata until the
+ * delegation is recalled.
+ *
+ * Recalls are issued via a callback function, provided by the application.
+ * Callback functions should act something like signal handlers.  You want to
+ * do as little as possible in the callback. Any major work should be deferred
+ * in some fashion as it's difficult to predict the context in which this
+ * function will be called.
+ *
+ * Once the delegation has been recalled, the application should return it as
+ * soon as possible. The application has client_deleg_timeout seconds to
+ * return it, after which the cmount structure is forcibly unmounted and
+ * further calls into it fail.
+ *
+ * The application can set the client_deleg_timeout config option to suit its
+ * needs, but it should take care to choose a value that allows it to avoid
+ * forcible eviction from the cluster in the event of an application bug.
```

Comment:
Ok, I'll need some guidance there. IIUC, the eviction timeout is supposed to be settable via a conf option, so in principle we should be able to just use ceph_conf_get to get the current value.

@jcsp : do you know how we would go about querying for this?

Reply:
MDSMap::get_session_timeout(). It's distributed in the MDSMap so that we know clients and MDSes aren't disagreeing on the value.
(Unfortunately, it's just a bare variable rather than wrapped up in some extensible config options thing we could use elsewhere, but such is life...)

Reply:
Hmm ok. I'm not sure I like the idea of shutting down on that -- it'd be nice if we could just reject changing that value to anything less than get_session_timeout returns.  The config observer doesn't really allow that -- the callbacks there are all void return.

The next best thing would be to just ignore any attempt to set it to an unreasonable value. Basically, limit the minimum timeout to what get_session_timeout returns, and probably log a message if we had to alter it.

The tricky bit with all of this is that we may not have the mdsmap yet when we start parsing the config. So, what we may need to do is:

1) have the handle_conf_change routine call get_session_timeout. If it's !=0 then we'll use that to cap the value of the conf change. If we end up altering it, we log a warning.

2) if get_session_timeout returns 0, then we don't know the timeout yet. Just accept what userland gave us.

3) when we get a new map (in Client::handle_mds_map), get the session timeout and vet it against the delegation timeout. Reduce the deleg timeout if necessary, and log a message if that occurs.

Reply:
I know you're writing all the NFS Ganesha code, but it seems like we really want a negotiation happening between Client and the CephFS FSAL so that both sides agree things are kosher before they proceed. If CephFS is silently setting a lower recall timeout, then Ganesha is going to be getting itself unexpectedly blacklisted, and that just sounds like a nightmare to debug.

Maybe discuss with @ajarr and the Manila people what they'd prefer to have happen in the case of misconfiguration?